clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -686,6 +696,207 @@ check_function_markers()
+        constexpr size_t RETURN_SZ = 3;
+        constexpr size_t SYSCALL_SZ = 2;
+
+        std::vector<memref_t> memrefs = {
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, BASE_PC, CALL_SZ),
+            // First signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, BASE_PC + CALL_SZ),
+            // Second signal. No intervening instr here. Should skip pre-signal instr
+            // check on return.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, SIG1_PC),
+            gen_instr(TID, SIG2_PC, INSTR_SZ),
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, SIG2_PC + INSTR_SZ, RETURN_SZ),
+            // sigreturn of the second signal.
+            gen_instr(TID, SIG2_PC + INSTR_SZ + RETURN_SZ, SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER,
+                       SIG2_PC + INSTR_SZ + RETURN_SZ + SYSCALL_SZ),
+            gen_instr(TID, SIG1_PC, INSTR_SZ),
+
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, SIG1_PC + INSTR_SZ, RETURN_SZ),
+            // sigreturn of the first signal.
+            gen_instr(TID, SIG1_PC + INSTR_SZ + RETURN_SZ, SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER,
+                       SIG1_PC + INSTR_SZ + RETURN_SZ + SYSCALL_SZ),
+            // Function marker of the call before the first signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle consecutive signals (that are nested at the same depth) without
+    // any intervening instr between them.
+    {
+        constexpr addr_t BASE_PC = 100;
+        constexpr addr_t SIG1_PC = 200;
+        constexpr addr_t SIG2_PC = 300;
+        constexpr size_t INSTR_SZ = 1;
+        constexpr size_t RETURN_SZ = 3;
+        constexpr size_t SYSCALL_SZ = 2;
+
+        std::vector<memref_t> memrefs = {
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, BASE_PC, CALL_SZ),
+            // First signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, BASE_PC + CALL_SZ),
+            gen_instr(TID, SIG1_PC, INSTR_SZ),
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, SIG1_PC + INSTR_SZ, RETURN_SZ),
+            // sigreturn.
+            gen_instr(TID, SIG1_PC + INSTR_SZ + RETURN_SZ, SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER,
+                       SIG1_PC + INSTR_SZ + RETURN_SZ + SYSCALL_SZ),
+
+            // Second signal with no intervening instr.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, SIG1_PC + INSTR_SZ),
```

Comment:
This is the wrong interruption PC: it should have the same interruption PC as the 1st signal.

This should fail the invariant checks -- or is that what the longstanding PR #5965 would check but it's not merged yet @lihasgupta ?

Reply:
I've fixed the interruption PC.